src/client/error.rs

```diff
@@ -0,0 +1,264 @@
+use bson;
+use byteorder;
+
+use client::coll::options::WriteModel;
+use std::{error, fmt, io, result, sync};
+
+/// A type for results generated by MongoDB related functions, where the Err type is mongodb::Error.
+pub type Result<T> = result::Result<T, Error>;
+
+/// The error type for MongoDB operations.
+#[derive(Debug)]
+pub enum Error {
+    IoError(io::Error),
+    EncoderError(bson::EncoderError),
+    DecoderError(bson::DecoderError),
+    WriteError(WriteException),
+    BulkWriteError(BulkWriteException),
+    ArgumentError(String),
+    OperationError(String),
+    ResponseError(String),
+    CursorMissingError,
+    LockError,
+    DefaultError(String),
+}
+
+/// The error type for Write-related MongoDB operations.
+#[derive(Debug)]
+pub struct WriteException {
+    kind: WriteExceptionKind,
+    code: i32,
+    message: String,
+}
+
+/// The two types of single-write related MongoDB operations.
+#[derive(Debug)]
+pub enum WriteExceptionKind {
+    WriteConcernError(bson::Document),
```

Comment:
A `writeConcernError` can also happen in multiple-write scenarios, for example, when executing write commands.